doc/sphinx/addendum/rewrite-rules.rst

```diff
@@ -0,0 +1,144 @@
+   startup.
+
+   .. exn:: Rewrite rule declaration requires enabling the flag "Allow Rewrite Rules".
+      :undocumented:
+
+Declaring symbols
+-----------------
+
+Rewrite rules operate on symbols, which are their own kind of constants.
+
+.. cmd:: {| Symbol | Symbols } {| {+ ( @assumpt ) } | @assumpt }
+   :name: Symbol; Symbols
+
+   Symbols are declared in the same way as axioms, by giving their type and using the keyword ``Symbol``.
+
+   .. coqtop:: all
+
+      Set Allow Rewrite Rules.
+      Symbol pplus : nat -> nat -> nat.
+      Notation "a ++ b" := (pplus a b).
+
+Declaring rules
+---------------
+
+Rewrite rules have two parts named pattern (lhs) and production (rhs).
+When a rule is applied, the term is matched against the pattern, subterms aligned with pattern holes are collected
+and substitued into the production which is returned.
+
+.. cmd:: Rewrite {| Rule | Rules } @ident := @rewrite_rule {* with @rewrite_rule }
+  :name: Rewrite; Rule; Rules
+
+   Rewrite rules are declared in blocks, which bear a name.
+
+  .. prodn::
+     rewrite_rule ::= @term ==> @term
+
+
+  .. coqtop:: all
+
+     Rewrite Rule pplus_rewrite :=
+          ?n ++ 0 ==> ?n
+     with ?n ++ S ?m ==> S (?n ++ ?m)
+     with 0 ++ ?n ==> ?n
+     with S ?n ++ ?m ==> S (?n ++ ?m).
+
+Pattern syntax
+--------------
+
+Patterns are read from a subpart of terms, with the following syntax :
+
+  .. prodn::
+     pat ::= @qualid {? @univ_annot }
+     | @pat {+ @arg_pat }
+     | @pat .( @qualid {? @univ_annot } )
+     | match @pat {? as @name in @qualid {* {| @name | _ } } } {? return @arg_pat } with {? %| } {*| @qualid {* {| @name | _ } } => @arg_pat } end
+     | fun ({+ @name } {? : @arg_pat}) => @arg_pat
+     | forall ({+ @name } {? : @arg_pat}), @arg_pat
+     arg_pat ::= ? @name | _ | @pat
+
+where :n:`@qualid {? @univ_annot }` can refer to symbols, inductives and constructors, but not arbitrary constants.
+
+Equivalently, we can see a pattern as a pair of a rigid head and a list of eliminations :
+
+  .. prodn::
+     head_pat ::= @qualid {? @univ_annot }
+     | fun ({+ @name } {? : @arg_pat}) => @arg_pat
+     | forall ({+ @name } {? : @arg_pat}), @arg_pat
+     elimination ::= <> {+ @pat }
+     | <> .( @qualid {? @univ_annot } )
+     | match <> {? as @name in @qualid {* {| @name | _ } } } {? return @arg_pat } with {*| @qualid {* {| @name | _ } } => @arg_pat } end
+     pat ::= @head_pat {* @elimination }
+     arg_pat ::= ? @name | _ | @pat
+
+To be a valid pattern, it also must have a symbol as its head pattern, since the mechanism to trigger rewrite rules needs a symbol.
+
+Productions
+-----------
+
+Productions are regular terms, which can also refer to matched holes in the pattern, through the :n:`? @name` syntax.
+
+
+Higher-order pattern holes
+--------------------------
+
+Pattern production lambda, prod and match introduce new variables in the context which need to be taken into account in the production.
```

Comment:
```suggestion
Pattern production lambda, prod and `match` introduce new variables in the context which need to be taken into account in the production.
```

Replace "lambda" with `fun` and whatever else it refers to.  No need to use terminology if a literal explanation is succinct and clear.  What is "prod"?  "Taken into account" meaning (mostly) picking a fresh name?

Reply:
I rephrased this sentence.
"Taking into account" is more than picking a fresh name, since the variables may be substituted why any term in the right-hand side